README.md
# math_ext

수식 확장 함수 모음

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 2.0.2
- vector_math 패키지 추가
- 두 좌표의 각도를 구하는 공식 수정

## 2.0.1
- 두 위경도 좌표 사이의 거리 공식 오류 수정

## 2.0.0
- extension 을 이용한 함수 형태 변경
- 패키지명 변경

## 1.1.0
- null safety migration

## 1.0.0
### 기본적인 기하연산 작성
**general math**
- clamp
- wrap
**geometry**
- radian <-> degree 변환
- radian/degree -> clock position 변환
- 두 벡터 사이의 투영 벡터 계산
- 두 벡터 수선 벡터 계산
**vincenty formula**
- 주어진 두 좌표사이의 거리 계산
- 주어진 좌표, 방위, 거리만큼 이동한 좌표 계산

lib/src/vector.dart
import 'dart:math';

import 'package:vector_math/vector_math.dart';

import 'angle.dart';

const double pi2 = pi * 2;
const double piDiv2 = pi / 2;

extension Vector on Point {
  num dot(Point other) => x * other.x + y * other.y;

  /// [point] 까지의 투영 벡터를 계산하여 반환.
  /// [offset]이 주어질 경우 더해서 반환.
  Point projectionVectorTo(
    Point point, [
    Point offset = const Point(0, 0),
  ]) =>
      offset + (point * (dot(point) / point.dot(point)));

  /// [point]에 내린 수선의 벡터를 계산하여 반환.
  /// [offset]이 주어질 경우 더해서 반환.
  Point perpendicularVectorTo(
    Point point, [
    Point offset = const Point(0, 0),
  ]) =>
      offset - (this - projectionVectorTo(point));

  /// from 좌표부터 to 좌표까지의 방위각을 라디안 단위로 반환.
  double azimuthRadianTo(Point to) {
    var startLongitudeRadians = radians(x.toDouble());
    var startLatitudeRadians = radians(y.toDouble());
    var endLongitudeRadians = radians(to.x.toDouble());
    var endLatitudeRadians = radians(to.y.toDouble());

    var dY = sin(endLongitudeRadians - startLongitudeRadians) *
        cos(endLatitudeRadians);
    var dX = cos(startLatitudeRadians) * sin(endLatitudeRadians) -
        sin(startLatitudeRadians) *
            cos(endLatitudeRadians) *
            cos(endLongitudeRadians - startLongitudeRadians);

    return atan2(dY, dX);
  }

  /// from 자표 부터 to 좌표 까지의 방위각을 도 단위로 반환.
  double azimuthDegreeTo(Point to) => toDegree(azimuthRadianTo(to));
}


lib/src/geometry.dart


part of geo_math;

// 평면상의 기하 연산.

const double pi2 = pi * 2;
const double piDiv2 = pi / 2;

/// [degree]를 라디안 단위로 변환.
double toRadian(double degree) => degree * (pi / 180);

/// [radian]을 도 단위로 변환.
double toDegree(double radian) => radian * (180 / pi);

/// 도 단위의 [azimuth]를 시계방향 각을 계산하여 시(1시, 2시) 방향 반환.
int degreeToClockPosition(double azimuth) {
  final angle = ((azimuth + 360) / 30 % 12).round();
  return angle == 0 ? 12 : angle;
}

/// 라디안 단위의 [azimuth]를 시계방향 각을 계산하여 시(1시, 2시) 방향 반환.
int radianToClockPosition(double azimuth) =>
    degreeToClockPosition(toDegree(azimuth));

/// 두 벡터[v1], [v2] 사이의 투영 벡터를 계산하여 반환.
/// [offset]이 주어질 경우 더해서 반환.
Vector2 getProjectionVectorBetween(Vector2 v1, Vector2 v2, {Vector2? offset}) =>
    (v2 * (v1.dot(v2) / v2.dot(v2))) + (offset ?? Vector2.zero());

/// 두 벡터[v1]에서 [v2]에 내린 수선의 벡터를 계산하여 반환.
/// [offset]이 주어질 경우 더해서 반환.
Vector2 getPerpendicularVectorBetween(Vector2 v1, Vector2 v2,
        {Vector2? offset}) =>
    ((v1 - getProjectionVectorBetween(v1, v2)) * -1) +
    (offset ?? Vector2.zero());


lib/src/geography.dart
import 'dart:math';

import 'angle.dart';
import 'vector.dart';

/// 지구상 두 점 사이의 거리 계산 및 한 점으로 부터 일정 거리 떨어진 점을 계산하는
/// Vincenty 공식 구현.
/// Based on WGS-84.

/// 지구 평편률
const double _flattening = 1 / 298.257223563;

/// 지구 중심부터 적도까지의 반지름(km)
const double _equatorialRadius = 6378137.0;

/// 지구 중심부터 북극 까지의 반지름(km)
const double _polarRadius = 6356752.314245;

/// 오차값 보정값
const double _epsilon = 1e-12;

extension GeoMathPoint on Point {
  /// [azimuth]방향으로 [distance]떨어진 점의 좌표를 계산하여 반환.
  Point offsetTo(double azimuth, double distance) {
    final sinAlpha1 = sin(azimuth);
    final cosAlpha1 = cos(azimuth);

    final tanU1 = (1 - _flattening) * tan(toRadian(y));
    final cosU1 = 1 / sqrt((1 + tanU1 * tanU1));
    final sinU1 = tanU1 * cosU1;

    final sigma1 = atan2(tanU1, cosAlpha1);
    final sinAlpha = cosU1 * sinAlpha1;
    final cosSqrAlpha = 1 - sinAlpha * sinAlpha;
    final uSqr = cosSqrAlpha *
        (_equatorialRadius * _equatorialRadius - _polarRadius * _polarRadius) /
        (_polarRadius * _polarRadius);

    final A =
        1 + (uSqr / 16384) * (4096 + uSqr * (-768 + uSqr * (320 - 175 * uSqr)));
    final B = (uSqr / 1024) * (256 + uSqr * (-128 + uSqr * (74 - 47 * uSqr)));

    var maxIteration = 200;
    var prevSigma = pi2;
    var sigma = distance / (_polarRadius * A);

    double cos2SigmaM;
    double sinSigma;
    double cosSigma;
    double deltaSigma;

    do {
      cos2SigmaM = cos(2 * sigma1 + sigma);
      sinSigma = sin(sigma);
      cosSigma = cos(sigma);
      deltaSigma = B *
          sinSigma *
          (cos2SigmaM +
              (B / 4) *
                  (cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM) -
                      (B / 6) *
                          cos2SigmaM *
                          (-3 + 4 * sinSigma * sinSigma) *
                          (-3 + 4 * cos2SigmaM * cos2SigmaM)));

      prevSigma = sigma;
      sigma = distance / (_polarRadius * A) + deltaSigma;
    } while ((sigma - prevSigma).abs() > _epsilon && --maxIteration > 0);

    if (maxIteration == 0) {
      throw StateError('Offset calculation faild to converge!');
    }

    final temp = sinU1 * sinSigma - cosU1 * cosSigma * cosAlpha1;
    final targetLat = atan2(sinU1 * cosSigma + cosU1 * sinSigma * cosAlpha1,
        (1 - _flattening) * sqrt(sinAlpha * sinAlpha + temp * temp));
    final lambda = atan2(
        sinSigma * sinAlpha1, cosU1 * cosSigma - sinU1 * sinSigma * cosAlpha1);
    final C = (_flattening / 16) *
        cosSqrAlpha *
        (4 + _flattening * (4 - 3 * cosSqrAlpha));
    final L = lambda -
        (1 - C) *
            _flattening *
            sinAlpha *
            (sigma +
                C *
                    sinSigma *
                    (cos2SigmaM +
                        C * cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM)));

    var targetLng = L + toRadian(x);

    if (targetLng > pi) {
      targetLng = targetLng - 2 * pi;
    }
    if (targetLng < -pi) {
      targetLng = targetLng + 2 * pi;
    }

    return Point(targetLng, targetLat);
  }

  /// 두 점 [p1]과 [p2] 사이의 거리를 미터 단위로 계산하여 반환.
  double metricDistanceTo(Point to) {
    final L = toRadian(to.x) - toRadian(x);
    final u1 = atan((1 - _flattening) * tan(toRadian(y)));
    final u2 = atan((1 - _flattening) * tan(toRadian(to.y)));
    final sinU1 = sin(u1), cosU1 = cos(u1);
    final sinU2 = sin(u2), cosU2 = cos(u2);

    double sinLambda,
        cosLambda,
        sinSigma,
        cosSigma,
        sigma,
        sinAlpha,
        cosSqAlpha,
        cos2SigmaM;
    double lambda = L, lambdaP;

    var maxIterations = 200;

    do {
      sinLambda = sin(lambda);
      cosLambda = cos(lambda);
      sinSigma = sqrt((cosU2 * sinLambda) * (cosU2 * sinLambda) +
          (cosU1 * sinU2 - sinU1 * cosU2 * cosLambda) *
              (cosU1 * sinU2 - sinU1 * cosU2 * cosLambda));

      if (sinSigma == 0) {
        return 0.0; // co-incident points
      }

      cosSigma = sinU1 * sinU2 + cosU1 * cosU2 * cosLambda;
      sigma = atan2(sinSigma, cosSigma);
      sinAlpha = cosU1 * cosU2 * sinLambda / sinSigma;
      cosSqAlpha = 1 - sinAlpha * sinAlpha;
      cos2SigmaM = cosSigma - 2 * sinU1 * sinU2 / cosSqAlpha;

      if (cos2SigmaM.isNaN) {
        cos2SigmaM = 0.0; // equatorial line: cosSqAlpha=0 (§6)
      }

      final C = _flattening /
          16 *
          cosSqAlpha *
          (4 + _flattening * (4 - 3 * cosSqAlpha));
      lambdaP = lambda;
      lambda = L +
          (1 - C) *
              _flattening *
              sinAlpha *
              (sigma +
                  C *
                      sinSigma *
                      (cos2SigmaM +
                          C * cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM)));
    } while ((lambda - lambdaP).abs() > 1e-12 && --maxIterations > 0);

    if (maxIterations == 0) {
      throw StateError('Distance calculation faild to converge!');
    }

    final uSq = cosSqAlpha *
        (_equatorialRadius * _equatorialRadius - _polarRadius * _polarRadius) /
        (_polarRadius * _polarRadius);
    final A = 1 + uSq / 16384 * (4096 + uSq * (-768 + uSq * (320 - 175 * uSq)));
    final B = uSq / 1024 * (256 + uSq * (-128 + uSq * (74 - 47 * uSq)));
    final deltaSigma = B *
        sinSigma *
        (cos2SigmaM +
            B /
                4 *
                (cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM) -
                    B /
                        6 *
                        cos2SigmaM *
                        (-3 + 4 * sinSigma * sinSigma) *
                        (-3 + 4 * cos2SigmaM * cos2SigmaM)));

    return _polarRadius * A * (sigma - deltaSigma);
  }
}


lib/src/vincenty_formula.dart


part of geo_math;

/// 지구상 두 점 사이의 거리 계산 및 한 점으로 부터 일정 거리 떨어진 점을 계산하는
/// Vincenty 공식 구현.
/// Based on WGS-84.

/// 지구 평편률
const double flattening = 1 / 298.257223563;

/// 지구 중심부터 적도까지의 반지름(km)
const double equatorialRadius = 6378137.0;

/// 지구 중심부터 북극 까지의 반지름(km)
const double polarRadius = 6356752.314245;

/// 오차값 보정값
const double epsilon = 1e-12;

/// [from]부터 [azimuth]방향으로 [distanceM]떨어진 점의 좌표를 계산하여 반환.
Map<String, double> calculateOffset(
  final double fromLat,
  final double fromLng,
  final double azimuth,
  final double distanceM,
) {
  if (fromLat == null ||
      fromLng == null ||
      azimuth == null ||
      distanceM == null) {
    throw ArgumentError.notNull();
  }

  final sinAlpha1 = sin(azimuth);
  final cosAlpha1 = cos(azimuth);

  final tanU1 = (1 - flattening) * tan(toRadian(fromLat));
  final cosU1 = 1 / sqrt((1 + tanU1 * tanU1));
  final sinU1 = tanU1 * cosU1;

  final sigma1 = atan2(tanU1, cosAlpha1);
  final sinAlpha = cosU1 * sinAlpha1;
  final cosSqrAlpha = 1 - sinAlpha * sinAlpha;
  final uSqr = cosSqrAlpha *
      (equatorialRadius * equatorialRadius - polarRadius * polarRadius) /
      (polarRadius * polarRadius);

  final A =
      1 + (uSqr / 16384) * (4096 + uSqr * (-768 + uSqr * (320 - 175 * uSqr)));
  final B = (uSqr / 1024) * (256 + uSqr * (-128 + uSqr * (74 - 47 * uSqr)));

  var maxIteration = 200;
  var prevSigma = pi2;
  var sigma = distanceM / (polarRadius * A);

  double cos2SigmaM;
  double sinSigma;
  double cosSigma;
  double deltaSigma;

  do {
    cos2SigmaM = cos(2 * sigma1 + sigma);
    sinSigma = sin(sigma);
    cosSigma = cos(sigma);
    deltaSigma = B *
        sinSigma *
        (cos2SigmaM +
            (B / 4) *
                (cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM) -
                    (B / 6) *
                        cos2SigmaM *
                        (-3 + 4 * sinSigma * sinSigma) *
                        (-3 + 4 * cos2SigmaM * cos2SigmaM)));

    prevSigma = sigma;
    sigma = distanceM / (polarRadius * A) + deltaSigma;
  } while ((sigma - prevSigma).abs() > epsilon && --maxIteration > 0);

  if (maxIteration == 0) {
    throw StateError('Offset calculation faild to converge!');
  }

  final temp = sinU1 * sinSigma - cosU1 * cosSigma * cosAlpha1;
  final targetLat = atan2(sinU1 * cosSigma + cosU1 * sinSigma * cosAlpha1,
      (1 - flattening) * sqrt(sinAlpha * sinAlpha + temp * temp));
  final lambda = atan2(
      sinSigma * sinAlpha1, cosU1 * cosSigma - sinU1 * sinSigma * cosAlpha1);
  final C = (flattening / 16) *
      cosSqrAlpha *
      (4 + flattening * (4 - 3 * cosSqrAlpha));
  final L = lambda -
      (1 - C) *
          flattening *
          sinAlpha *
          (sigma +
              C *
                  sinSigma *
                  (cos2SigmaM +
                      C * cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM)));

  var targetLng = L + toRadian(fromLng);

  if (targetLng > pi) {
    targetLng = targetLng - 2 * pi;
  }
  if (targetLng < -pi) {
    targetLng = targetLng + 2 * pi;
  }

  return {'latitude': toDegree(targetLat), 'longitude': toDegree(targetLng)};
}

/// 두 점 [p1]과 [p2] 사이의 거리를 미터 단위로 계산하여 반환.
double calculateDistance(
  final double fromLat,
  final double fromLng,
  final double toLat,
  final double toLng,
) {
  if (fromLat == null || fromLng == null || toLat == null || toLng == null) {
    throw ArgumentError.notNull();
  }

  final L = toRadian(toLng) - toRadian(fromLng);
  final u1 = atan((1 - flattening) * tan(toRadian(fromLat)));
  final u2 = atan((1 - flattening) * tan(toRadian(toLat)));
  final sinU1 = sin(u1), cosU1 = cos(u1);
  final sinU2 = sin(u2), cosU2 = cos(u2);

  double sinLambda,
      cosLambda,
      sinSigma,
      cosSigma,
      sigma,
      sinAlpha,
      cosSqAlpha,
      cos2SigmaM;
  double lambda = L, lambdaP;

  var maxIterations = 200;

  do {
    sinLambda = sin(lambda);
    cosLambda = cos(lambda);
    sinSigma = sqrt((cosU2 * sinLambda) * (cosU2 * sinLambda) +
        (cosU1 * sinU2 - sinU1 * cosU2 * cosLambda) *
            (cosU1 * sinU2 - sinU1 * cosU2 * cosLambda));

    if (sinSigma == 0) {
      return 0.0; // co-incident points
    }

    cosSigma = sinU1 * sinU2 + cosU1 * cosU2 * cosLambda;
    sigma = atan2(sinSigma, cosSigma);
    sinAlpha = cosU1 * cosU2 * sinLambda / sinSigma;
    cosSqAlpha = 1 - sinAlpha * sinAlpha;
    cos2SigmaM = cosSigma - 2 * sinU1 * sinU2 / cosSqAlpha;

    if (cos2SigmaM.isNaN) {
      cos2SigmaM = 0.0; // equatorial line: cosSqAlpha=0 (§6)
    }

    final C =
        flattening / 16 * cosSqAlpha * (4 + flattening * (4 - 3 * cosSqAlpha));
    lambdaP = lambda;
    lambda = L +
        (1 - C) *
            flattening *
            sinAlpha *
            (sigma +
                C *
                    sinSigma *
                    (cos2SigmaM +
                        C * cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM)));
  } while ((lambda - lambdaP).abs() > 1e-12 && --maxIterations > 0);

  if (maxIterations == 0) {
    throw StateError('Distance calculation faild to converge!');
  }

  final uSq = cosSqAlpha *
      (equatorialRadius * equatorialRadius - polarRadius * polarRadius) /
      (polarRadius * polarRadius);
  final A = 1 + uSq / 16384 * (4096 + uSq * (-768 + uSq * (320 - 175 * uSq)));
  final B = uSq / 1024 * (256 + uSq * (-128 + uSq * (74 - 47 * uSq)));
  final deltaSigma = B *
      sinSigma *
      (cos2SigmaM +
          B /
              4 *
              (cosSigma * (-1 + 2 * cos2SigmaM * cos2SigmaM) -
                  B /
                      6 *
                      cos2SigmaM *
                      (-3 + 4 * sinSigma * sinSigma) *
                      (-3 + 4 * cos2SigmaM * cos2SigmaM)));

  return polarRadius * A * (sigma - deltaSigma);
}
